docs/installation/database/passing-configuration.rst

```diff
@@ -0,0 +1,430 @@
+are described in :doc:`configuration`.
+
+The Open Data Cube uses configuration files and/or environment variables to determine how to connect to databases.
+
+Overview
+========
+
+The normal flow of configuration processing can be bypassed by `explicitly passing`_ in a complete configuration.
+Such explicit configuration blocks reading of configuration from environment variables.
+
+Otherwise, configuration is read from a `configuration file`_ and/or `environment variables`_.
+
+One configuration can define multiple environments, so users must `choose one`_.
+
+The configuration engine in 1.9 is not 100% compatible with the previous configuration engine.  Advanced
+users and developers upgrading 1.8 systems should read the `migration notes`_.
+
+.. _`explicitly passing`: #Explicit-configurations
+.. _`configuration file`: #File-configurations
+.. _`choose one`: #The-Active-Environment
+.. _`environment variables`: #Generic-Environment-Variable-Overrides
+.. _`migration notes`: #Migrating-from-datacube-1.8
+
+1. Explicit configuration
+-------------------------
+
+Configuration can be passed in explicitly, without ever reading from a configuration file on disk.
+
+When explicit configuration is passed in, it takes precedence over configuration by environment variable.
+
+1a. Via Python (str or dict)
+++++++++++++++++++++++++++++
+
+A valid configuration dictionary can be passed in directly to the
+:py:class:`Datacube` constructor with the ``raw_config`` argument, without
+serialising to a string:
+
+.. code-block:: python
+
+   dc = Datacube(raw_config={
+      "default": {
+         "index_driver": "postgres",
+         "db_url": "postgresql:///mydb"
+      }
+   })
+
+The ``raw_config`` argument can also be passed config as a string, in either INI or YAML format:
+
+.. code-block:: python
+
+   dc = Datacube(raw_config="""
+   default:
+     # Connect to database mydb over local socket with OS authentication.
+     db_url: postgresql:///mydb
+   """)
+
+1b. As a string, via the datacube CLI
++++++++++++++++++++++++++++++++++++++
+
+The contents of a configuration file can be passed into the ``datacube`` CLI via the ``-R`` or
+``--raw-config`` command line option:
+
+::
+
+   datacube --raw-config "default: {db_database: this_db}"
+
+Output from a script that generates a configuration file dynamically can be passed in using
+a BASH backquote string:
+
+::
+
+   datacube --raw-config "`config_file_generator --option blah`"
+
+1c. As a string, via an Environment Variable
+++++++++++++++++++++++++++++++++++++++++++++
+
+If raw configuration has not been passed in explicitly via methods 1a. or 1b.
+above, then the contents of a configuration file can be written in full to the
+:envvar:`ODC_CONFIG` environment variable:
+
+.. code-block:: console
+
+   $ ODC_CONFIG="default: {db_database: this_db}"
+   $ datacube check    # will use the this_db database
+
+
+2. File configuration
```

Comment:
These sections seem to be sorted according to the code flow, but I would intuitively expect them to be sorted by commonality so the first subsection is what applies to most users. Who is the target audience for this document?

Reply:
The `configuration.rst` (Configuration basics) document will hopefully satisfy most users.  This is really for more advanced users including developers.

Reply:
I would suggest ordering the sections by popularity so people can stop reading the document after they have found the section that applies to their scenario, so if most users configure through a file, that section should come first. The a-b-c-d subsections are short so they can stay as they are.

If reordering the sections leads to terms being defined several sections after they are used, well, then it's perhaps better to keep the current order.

Reply:
Hang on I think I see what you mean.